<?xml version="1.0"?>
<robot name="rapharobot">
  <gazebo>
    <plugin name="gazebo_ros2_control"
    filename ="libgazebo_ros2_control.so">
    <parameters >$(find gpg_urdf )/controllers.yaml </parameters>
    </plugin >
</gazebo> 

<gazebo reference="gazebo_camera_link">
  <sensor type="camera" name="camera">

  <update_rate>30</update_rate>
    <camera name="camera_name">
  
    <horizontal_fov>0.915</horizontal_fov>
    <image>
        <width>320</width>
        <height>240</height>
        <format>R8G8B8</format>
    </image>
    </camera>
    
    <always_on>1</always_on>
    <update_rate>30</update_rate>
    <visualize>true</visualize>
    
    <plugin name="libgazebo_ros_camera"
    filename ="libgazebo_ros_camera.so">
        <ros >
           <remapping>image_raw:=image</remapping>
           <remapping>camera_info:=camera_info</remapping>
       </ros>
   
       <camera_name>camera_name2</camera_name>
       <frame_name>camera_link</frame_name>
   
       <imageTopicName>image_raw</imageTopicName>
       <cameraInfoTopicName>camera_info></cameraInfoTopicName>
   
       <distortionk1>0.080938</distortionk1>
       <distortionk2>-0.257429</distortionk2>
       <distortionk3>0.00000</distortionk3>
       <distortionT1>-0.002955</distortionT1>
       <distortionT2>0.003629</distortionT2>
  </plugin>
    
</sensor>
</gazebo>
  

  
<ros2_control name="gpg_remote" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    
    <joint name="left_wheel">
      <command_interface name="velocity"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>  
    </joint>
    <joint name="right_wheel">
      <command_interface name="velocity"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>  
    </joint>
    <joint name="servo">
      <command_interface name="position"/>
      <state_interface name="position"/>
    </joint>
    <sensor name="line">
       <state_interface name="level0"/>
       <state_interface name="level1"/>
       <state_interface name="level2"/>
       <state_interface name="level3"/>
       <state_interface name="level4"/>
    </sensor>
    <sensor name="battery">
       <state_interface name="voltage"/>
    </sensor>
  </ros2_control>

  <material name="blue">
    <color rgba="0 0 0.8 1"/>
  </material>

  <material name="white">
    <color rgba="1 1 1 1"/>
  </material>

  <link name="base_link">
    <visual>
      <geometry>
        <box size="0.11 0.10 0.07" />
      </geometry>
      <origin xyz="0 0 0.07"/>
      <material name="blue"/>
    </visual>
  <collision>
    <geometry>
      <box size="0.11 0.10 0.07" />
    </geometry>
      <origin xyz="0 0 0.07"/>
  </collision>
  <inertial>
    <mass value="0.1"/>
    <inertia ixx="1e-3" ixy="0.0" ixz="0.0" iyy="1e-3" iyz="0.0" izz="1e-3"/>
  </inertial>
  </link>

  <link name="link_right_wheel">
    <visual>
      <geometry>
        <cylinder length="0.05" radius="0.0325"/>
      </geometry>
      <origin rpy="1.57 0 0"/>
      <material name="white"/>
    </visual>
      <collision>
    <geometry>
      <cylinder length="0.05" radius="0.0325"/>
    </geometry>
      <origin rpy="1.57 0 0"/>
  </collision>
  <inertial>
    <mass value="1"/>
    <inertia ixx="1e-3" ixy="0.0" ixz="0.0" iyy="1e-3" iyz="0.0" izz="1e-3"/>
  </inertial>
  </link>

  <joint name="right_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="link_right_wheel"/>
    <axis xyz="0 1 0"/>
    <origin xyz="0 -0.07 0.0325" />
  </joint>

  <link name="link_left_wheel">
    <visual>
      <geometry>
        <cylinder length="0.05" radius="0.0325"/>
      </geometry>
       <origin rpy="1.57 0 0"/>
      <material name="white"/>
    </visual>
      <collision>
    <geometry>
      <cylinder length="0.05" radius="0.0325"/>
    </geometry>
       <origin rpy="1.57 0 0"/>
  </collision>
  <inertial>
    <mass value="1"/>
    <inertia ixx="1e-3" ixy="0.0" ixz="0.0" iyy="1e-3" iyz="0.0" izz="1e-3"/>
  </inertial>
  </link>

  <joint name="left_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="link_left_wheel"/>
    <origin xyz="0 0.07 0.0325" />
    <axis xyz="0 1 0"/>
  </joint>


  <link name="servo_link">
    <visual>
      <geometry>
        <box size="0.02 0.01 0.03" />
      </geometry>
       <origin rpy="0.06 0 0"/>
      <material name="white"/>
    </visual>
      <collision>
    <geometry>
      <box size="0.02 0.01 0.03" />
    </geometry>
  </collision>
  <inertial>
    <mass value="0.1"/>
    <inertia ixx="1e-6" ixy="0.0" ixz="0.0" iyy="1e-6" iyz="0.0" izz="1e-6"/>
  </inertial>
  </link>

  <joint name="servo" type="continuous">
    <parent link="base_link"/>
    <child link="servo_link"/>
    <origin xyz="0.09 0 0.07" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>


  <link name="camera_link">
    <visual>
      <geometry>
        <box size="0.03 0.03 0.001" />
      </geometry>
       <origin rpy="0.06 0 0"/>
      <material name="white"/>
    </visual>
  </link>

  <link name="gazebo_camera_link">
  </link>

  <joint name="gazebo_camera" type="fixed">
    <parent link="servo_link"/>
    <child link="gazebo_camera_link"/>
    <origin xyz="0.018 0 0.045"/>
  </joint>

  <joint name="camera" type="fixed">
    <parent link="servo_link"/>
    <child link="camera_link"/>
    <origin xyz="0.018 0 0.045" rpy="1.57 3.14 1.57"/>
  </joint>


  <link name="suporte_link">
    <visual>
       <geometry>
          <sphere radius="0.02"/>
       </geometry>
          <origin rpy="0.06 0.00 0.00"/>
          <material name="white"/>
     </visual>
     <collision>
     
       <geometry>
           <sphere radius="0.02"/>
       </geometry>
  
    </collision>
    
    <inertial>
        <mass value="2"/>
        <inertia ixx="0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
    </inertial>
  </link>

  <gazebo reference="suporte_link">
    <mu1>0</mu1>
    <mu2>0</mu2>
  </gazebo>

   
 <joint name="suporte" type="continuous">
    <parent link="base_link"/>
    <child link="suporte_link"/>
    <origin xyz="-0.20 0 0.0203" rpy="0.00 0.00 0.00"/>
  </joint>
  
  
  
  
  
  
</robot>
